docs/adr/adr-009-telemetry.md

```diff
@@ -0,0 +1,220 @@
+decision for the celestia-node team is to cover with traces only the _necessary_ for the current "ShrEx" stack code as 
+the initial response to the ADR, leaving the rest to be integrated in the background for the devs in the team once they 
+are free as well as for the efficient bootstrapping into the code for the new devs.
+
+The next biggest priority - incentivized Testnet can be largely covered with traces as well. All participant will submit
+traces from their nodes to any provided backend endpoint by us during the whole network lifespan. Later on, we will be 
+able to verify the data of each participant by querying historical traces. This is the feature that some backend solutions 
+provide, which we can use as well to extract valuable on how the network performs in macro view.
+
+### Tooling/Dependencies
+
+#### Golang API/Shim
+
+The decision is to use [opentelemetry-go](https://github.com/open-telemetry/opentelemetry-go) for both Metrics and Tracing:
+* Minimal and golang savvy API/shim which gathers years of experience from OpenCensus/OpenMetrics and [CNCF](https://www.cncf.io/)
+* Backends/exporters for all the existing timeseries monitoring DBs, e.g. Prometheus, InfluxDB. As well as tracing backends
+  Jaeger, Uptrace, etc.
+* https://github.com/uptrace/opentelemetry-go-extra/tree/main/otelzap with logging engine with use - Zap
+
+The discussion over this decision can be found in [celestia-node#663](https://github.com/celestiaorg/celestia-node/issues/663)
+and props to @liamsi for initial kickoff and a deep dive into OpenTelemetry.
+
+#### Tracing Backends
+
+For the tracing there are 3 modern OSS tools are recommended. All of them have bidirectional support with OpenTelemetry:
+* [Uptrace](https://get.uptrace.dev/guide/#what-is-uptrace)
+  * The most recent (~1 year)
+  * The richest UI
+  * Tight to Clickhouse DB
+  * Made by OpenTelemetry
+  * The most lightweight
+* [Jaeger](https://www.jaegertracing.io/)
+  * The most mature
+  * Started by Uber, now supported by CNCF
+  * Supports multiple storages(ScyllaDB, InfluxDB, Amazon DynamoDB)
+* [Graphana Tempo](https://grafana.com/oss/tempo/)
+  * Deep integration with Graphana/Prometheus
+  * Relatively new (~2 years)
+  * Uses Azure, GCS, S3 or local disk for storage
+
+Each of these backends can be used independently and depending on the use case. For us, these are main use cases: 
+* Local development/debugging for the private network or even public network setup
+> I am personally planning to set up the lightweight Uptrace for the local light node. Just to play around and observe 
+> things
+* Data collection from the Testground test runs
+* Bootstrappers monitoring infrastructure
+* Data collection from the incentivized testnet participants
+
+There is no strict decision on which of these backends and where to use. People taking ownership of any listed vectors
+are free to use any recommended solution or any unlisted.
+
+#### Metrics Backend
+
+// WIP
```

Comment:
Yes

Reply:
> I think this is equally important
Initially, I also thought the same a made a deep dive into metrics, but after doing a deeper dive into tracing and how it is different, I disagree with metrics being important for us in the current state. Now, tracing have a higher priority:
* Easier to integrate. I will try to add here a comparison section 
* Better match for the current requirements. Tracing alone solves both reconstruction and incentivized testnet, while metrics don't.
* Metrics are good to measure things and analyzing performance, finding averages, and providing histograms of isolated processes. Tracing allows you to analyze any operation as the tree of dependencies seeing each step. I feel like I am repeating myself now.

Reply:
> Better match for the current requirements. Tracing alone solves both reconstruction and incentivized testnet, while metrics don't.

Tracing might be too granular as far I understand. While it "solves" tracing and then optimizing e.g. reconstruction it does not "solve" the incentivized testnet as we certainly want less granular metrics like how long requests took, or resource consumption (e.g. mem/CPU usage) over time. Metrics can be useful for both: a detailed graphana dashboard and potenially also for the incentrivized testnet dashboard.

> Tracing allows you to analyze any operation as the tree of dependencies seeing each step.

Yes, and that is great but we also need less granular metrics to consume directly IMO. We can then correlate them to traces but only if we have them. e.g. memory or CPU spikes. or an increase in the number of incoming share requests, a spike or decrease in the number of connected peers and so forth.

Reply:
>  I will try to add here a comparison section

I think a short section about this might help to spread awareness for the team what the diff is.

Reply:
> econstruction it does not "solve" the incentivized testnet as we certainly want less granular metrics like how long requests took, or resource consumption (e.g. mem/CPU usage) over time. 

@liamsi, According to the mock figma board, there won't be any resource consumption related metrics. How long a request is a tracing-related problem. _We still didn't mention at least one metric that is required for incentivized testnet and cannot be solved via metrics._ 

The only one I could think of now is the uptime, but we don't really care about it as long as the node sends 
us traces.

Reply:
However though, we will have enought time for metrics until incentivised testnet anyway, according to current quarterly roadmap. In the worst case scenario, we will still be able to fulfill incentivized testnet needs with tracing, but it is very unlikely that we will have to.

Reply:
> According to the mock figma board, there won't be any resource consumption related metrics. 

Yes, this is less about the figma board. Instead, or additionally, we want a more complete telemetry dashboard. It will display some of the traces and I think uptime, CPU and disk usage usage (per node and "global"), and similarly networking IO etc. I do not have an exhaustive list in mind but these are the bare minimal set of metrics I can think of. I am sure we can have sth more specific to the celestia-node software as well. But would need to think about this more. I assume that the node team will do a deep dive into the possible metrics and come up with a more complete list.